File: clients/typescript/src/middleware.ts
// Resource-server middleware for Express and Fastify. A backend becomes a
// Legant-protected resource server in a few lines: verify the Bearer token, attach
// the verified Claims to the request, and (optionally) authorize a per-request
// Action. This is the delegation-aware analog of a generic OIDC JWT middleware — it
// understands the act chain, the constraint dimensions, RFC 8707 audience
// canonicalization, and the signed revocation feed.
//
// Express and Fastify are NOT dependencies of this package; the request/response
// objects are typed structurally so importing this module pulls in nothing.

import { Verifier, Claims, type Action } from './verifier.js';

/** Extracts the token from an `Authorization: Bearer <token>` header value. */
export function bearerToken(authHeader?: string): string {
  if (!authHeader) throw new Error('missing Authorization header');
  const m = /^Bearer\s+(.+)$/i.exec(authHeader.trim());
  if (!m) throw new Error('Authorization header is not a Bearer token');
  return m[1].trim();
}

/** Verifies the Bearer token from an Authorization header value (throws on failure). */
export function authenticate(verifier: Verifier, authHeader?: string): Claims {
  return verifier.verify(bearerToken(authHeader));
}

// Minimal structural shapes so we depend on neither express nor fastify types.
interface ReqLike {
  headers: Record<string, string | string[] | undefined>;
  legant?: Claims;
}
interface ExpressResLike {
  status(code: number): ExpressResLike;
  set(field: string, value: string): ExpressResLike;
  send(body: string): unknown;
}
type ExpressNext = (err?: unknown) => void;
interface FastifyReplyLike {
  code(statusCode: number): FastifyReplyLike;
  header(name: string, value: string): FastifyReplyLike;
  send(body: string): unknown;
}

function authHeaderOf(req: ReqLike): string | undefined {
  const h = req.headers['authorization'] ?? req.headers['Authorization'];
  return Array.isArray(h) ? h[0] : h;
}

// ---- Express ---------------------------------------------------------------

/**
 * Express middleware that verifies the Bearer token and stores the Claims on
 * `req.legant`. On failure it sends 401 with an RFC 6750 challenge. Revocation
 * behavior follows how the Verifier was built (a revocation feed → revoked tokens
 * 401 here; otherwise revocation is bounded by the token TTL).
 *
 *   app.use(expressAuth(verifier));
 *   app.get('/data', expressRequireScope('data:read'), handler);
 */
export function expressAuth(verifier: Verifier) {
  return (req: ReqLike, res: ExpressResLike, next: ExpressNext): void => {
    try {
      req.legant = authenticate(verifier, authHeaderOf(req));
      next();
    } catch (e) {
      expressChallenge(res, 401, 'invalid_token', errMsg(e));
    }
  };
}

/** Express middleware (mount AFTER expressAuth) requiring the token to carry scope. */
export function expressRequireScope(scope: string) {
  return expressRequireAction(() => ({ scope }));
}

/** Express middleware (mount AFTER expressAuth) authorizing a per-request Action. */
export function expressRequireAction(action: (req: ReqLike) => Action) {
  return (req: ReqLike, res: ExpressResLike, next: ExpressNext): void => {
    const claims = req.legant;
    if (!claims) return void expressChallenge(res, 401, 'invalid_token', 'no verified token (mount expressAuth first)');
    try {
      claims.authorize(action(req));
      next();
    } catch (e) {
      expressChallenge(res, 403, 'insufficient_scope', errMsg(e));
    }
  };
}

function expressChallenge(res: ExpressResLike, code: number, err: string, desc: string): void {
  res
    .status(code)
    .set('WWW-Authenticate', `Bearer error="${err}", error_description="${desc}"`)
    .send(`${err}: ${desc}`);
}

// ---- Fastify ---------------------------------------------------------------

/**
 * Fastify preHandler hook that verifies the Bearer token and stores the Claims on
 * `request.legant`.
 *
 *   fastify.addHook('preHandler', fastifyAuth(verifier));
 *   fastify.get('/data', { preHandler: fastifyRequireScope('data:read') }, handler);
 */
export function fastifyAuth(verifier: Verifier) {
  return async (req: ReqLike, reply: FastifyReplyLike): Promise<void> => {
    try {
      req.legant = authenticate(verifier, authHeaderOf(req));
    } catch (e) {
      fastifyChallenge(reply, 401, 'invalid_token', errMsg(e));
    }
  };
}

/** Fastify preHandler (after fastifyAuth) requiring the token to carry scope. */
export function fastifyRequireScope(scope: string) {
  return fastifyRequireAction(() => ({ scope }));
}

/** Fastify preHandler (after fastifyAuth) authorizing a per-request Action. */
export function fastifyRequireAction(action: (req: ReqLike) => Action) {
  return async (req: ReqLike, reply: FastifyReplyLike): Promise<void> => {
    const claims = req.legant;
    if (!claims) return void fastifyChallenge(reply, 401, 'invalid_token', 'no verified token (add fastifyAuth first)');
    try {
      claims.authorize(action(req));
    } catch (e) {
      fastifyChallenge(reply, 403, 'insufficient_scope', errMsg(e));
    }
  };
}

function fastifyChallenge(reply: FastifyReplyLike, code: number, err: string, desc: string): void {
  reply
    .code(code)
    .header('WWW-Authenticate', `Bearer error="${err}", error_description="${desc}"`)
    .send(`${err}: ${desc}`);
}

// ---- self-hosted MCP server helper -----------------------------------------

/**
 * Extracts the tool name from a JSON-RPC MCP `tools/call` request body, for
 * resource servers that ARE an MCP server. Pair with claims.authorize:
 *
 *   const name = mcpToolName(req.body);
 *   req.legant!.authorize({ scope: toolScopes[name], tool: name });
 */
export function mcpToolName(body: unknown): string {
  const b = typeof body === 'string' ? JSON.parse(body) : (body as { method?: string; params?: { name?: string } });
  if (!b || b.method !== 'tools/call') throw new Error(`not a tools/call (method=${(b as { method?: string })?.method})`);
  const name = b.params?.name;
  if (!name) throw new Error('tools/call missing params.name');
  return name;
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}

File: sdk/middleware.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

// This file is the resource-server INTEGRATION layer: drop-in net/http (and
// chi-compatible) middleware so a backend becomes a Legant-protected resource
// server in a few lines, instead of wiring Verify/Authorize by hand. It is the
// delegation-aware analog of a generic OIDC JWT middleware — it understands the
// act delegation chain, the constraint dimensions, RFC 8707 audience
// canonicalization, and the signed revocation feed, none of which a plain
// access-token middleware models.

type ctxKey int

const claimsCtxKey ctxKey = 0

// ClaimsFrom returns the verified Claims stored by Authenticate, if any.
func ClaimsFrom(ctx context.Context) (*Claims, bool) {
	c, ok := ctx.Value(claimsCtxKey).(*Claims)
	return c, ok
}

// MustClaims returns the verified Claims or panics — use only inside handlers
// mounted behind Authenticate (where the claims are guaranteed present).
func MustClaims(ctx context.Context) *Claims {
	c, ok := ClaimsFrom(ctx)
	if !ok {
		panic("sdk: no Legant claims in context (is Authenticate mounted?)")
	}
	return c
}

// Authenticate verifies the request's Bearer token with v and, on success, stores
// the Claims in the request context for downstream handlers (read them with
// ClaimsFrom). On any verification failure it writes 401 with an RFC 6750
// WWW-Authenticate challenge and does NOT call the next handler. It does not
// authorize a specific action — compose RequireScope / RequireAction for that, or
// call Claims.Authorize inside the handler.
//
// Revocation behavior follows how v was built: WithRevocationFeed makes a revoked
// token 401 here; WithFeedFailClosed couples availability to the feed. Without a
// feed, revocation is bounded by the token's TTL. Choose deliberately.
func Authenticate(v *Verifier) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			tok, err := bearer(r)
			if err != nil {
				challenge(w, http.StatusUnauthorized, "invalid_request", err.Error())
				return
			}
			claims, err := v.Verify(tok)
			if err != nil {
				challenge(w, http.StatusUnauthorized, "invalid_token", err.Error())
				return
			}
			next.ServeHTTP(w, r.WithContext(context.WithValue(r.Context(), claimsCtxKey, claims)))
		})
	}
}

// RequireScope authorizes that the request's token carries scope. Mount it AFTER
// Authenticate. For per-request constraints (amount/resource/tool/time), use
// RequireAction instead, or call Claims.Authorize in the handler.
func RequireScope(scope string) func(http.Handler) http.Handler {
	return RequireAction(func(*http.Request) Action { return Action{Scope: scope} })
}

// RequireAction derives an Action from each request and authorizes it against the
// verified token, writing 403 on denial. Mount it AFTER Authenticate.
//
//	r.With(sdk.Authenticate(v), sdk.RequireAction(func(r *http.Request) sdk.Action {
//	    return sdk.Action{Scope: "warehouse:query", Resource: r.URL.Query().Get("schema")}
//	})).Get("/query", handler)
func RequireAction(action func(*http.Request) Action) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			claims, ok := ClaimsFrom(r.Context())
			if !ok {
				challenge(w, http.StatusUnauthorized, "invalid_token", "no verified token (mount Authenticate first)")
				return
			}
			if err := claims.Authorize(action(r)); err != nil {
				challenge(w, http.StatusForbidden, "insufficient_scope", err.Error())
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// bearer extracts the token from an Authorization: Bearer header.
func bearer(r *http.Request) (string, error) {
	h := r.Header.Get("Authorization")
	if h == "" {
		return "", fmt.Errorf("missing Authorization header")
	}
	const p = "Bearer "
	if len(h) <= len(p) || !strings.EqualFold(h[:len(p)], p) {
		return "", fmt.Errorf("Authorization header is not a Bearer token")
	}
	return strings.TrimSpace(h[len(p):]), nil
}

// challenge writes an RFC 6750 error response.
func challenge(w http.ResponseWriter, status int, code, desc string) {
	w.Header().Set("WWW-Authenticate", fmt.Sprintf(`Bearer error=%q, error_description=%q`, code, desc))
	http.Error(w, code+": "+desc, status)
}

// ---- self-hosted MCP server helper ----------------------------------------

// MCPToolName extracts the tool name from a JSON-RPC MCP "tools/call" request
// body, for resource servers that ARE an MCP server (rather than sitting behind
// the gateway). Pair it with Claims.Authorize to gate a tool before dispatch:
//
//	name, _ := sdk.MCPToolName(body)
//	if err := claims.Authorize(sdk.Action{Scope: toolScopes[name], Tool: name}); err != nil { ... }
func MCPToolName(body []byte) (string, error) {
	var req struct {
		Method string `json:"method"`
		Params struct {
			Name string `json:"name"`
		} `json:"params"`
	}
	if err := json.Unmarshal(body, &req); err != nil {
		return "", fmt.Errorf("not a JSON-RPC request: %w", err)
	}
	if req.Method != "tools/call" {
		return "", fmt.Errorf("not a tools/call (method=%q)", req.Method)
	}
	if req.Params.Name == "" {
		return "", fmt.Errorf("tools/call missing params.name")
	}
	return req.Params.Name, nil
}
